File: cap-1/src/verify.mjs
#!/usr/bin/env node
// ─────────────────────────────────────────────────────────────────────────────
// CAP-1 CONFORMANCE VERIFIER
//
// Node built-ins only. No dependencies, no installer, no network, no clock.
// Runs anywhere Node runs, including inside a boundary that admits no packages.
//
// It checks the eight normative rules of the Coverage Attestation Profile.
// Every rule exists because its absence is a way a coverage claim can be false
// while looking complete. Each refusal names the rule and the evidence.
// ─────────────────────────────────────────────────────────────────────────────
import { readFileSync } from 'node:fs';

const DISPOSITIONS = new Set(['not_applicable','disabled_by_policy','unsupported_input',
  'resource_exhausted','failed','unavailable','out_of_scope','withheld']);
const BASIS = new Set(['catalogue','enumeration','declared']);
const HEX = /^[0-9a-f]{32,128}$/;

export function verify(doc) {
  const fail = [];
  const F = (rule, grounds, evidence) => fail.push({ rule, grounds, evidence });

  // R0 — shape. A document that is not a CAP-1 document is refused, not coerced.
  if (!doc || typeof doc !== 'object') return { ok:false, failures:[{rule:'R0-shape',grounds:'not an object'}] };
  if (doc.profile !== 'cap/1') F('R0-shape','profile is not cap/1',{profile:doc.profile ?? null});
  if (!doc.subject || typeof doc.subject.ref !== 'string') F('R0-shape','subject.ref absent',{});
  if (!Array.isArray(doc.strata) || doc.strata.length === 0) F('R0-shape','no strata',{});
  if (!doc.integrity || typeof doc.integrity.complete !== 'boolean') F('R0-shape','integrity.complete absent',{});
  if (fail.length) return { ok:false, failures:fail };

  const ids = new Set();
  for (const s of doc.strata) {
    const at = `strata[${s.id ?? '?'}]`;

    if (typeof s.id !== 'string' || ids.has(s.id)) F('R0-shape','stratum id absent or duplicated',{at});
    ids.add(s.id);

    // R1 — NO SILENT REMAINDER.
    // eligible must equal examined plus every individually accounted unexamined unit.
    // A remainder that reconciles only by arithmetic is a population of units whose
    // fate is unrecorded, and unrecorded units are exactly where a negative hides.
    const un = Array.isArray(s.unexamined) ? s.unexamined : null;
    if (!un) { F('R1-no-silent-remainder','unexamined is not an array',{at}); continue; }
    const sum = s.examined + un.length;
    if (sum !== s.eligible)
      F('R1-no-silent-remainder','eligible does not equal examined plus accounted unexamined',
        {at, eligible:s.eligible, examined:s.examined, accounted:un.length, remainder:s.eligible - sum});

    // R2 — CLOSED DISPOSITION VOCABULARY.
    // An open reason string is where an uncomfortable disposition goes to hide.
    for (const u of un) {
      if (!DISPOSITIONS.has(u.disposition))
        F('R2-closed-disposition','disposition outside the closed vocabulary',{at, unit:u.unit, disposition:u.disposition});
      if (typeof u.unit !== 'string' || !u.unit.length)
        F('R2-closed-disposition','unexamined entry names no unit',{at});

      // R3 — WITHHOLDING IS DIGEST-BOUND.
      // A withheld unit that carries no digest of what was withheld cannot later be
      // shown to be the unit withheld rather than one substituted afterwards.
      if (u.disposition === 'withheld' && !HEX.test(String(u.withheld_digest ?? '')))
        F('R3-withholding-digest-bound','withheld unit carries no digest of the withheld value',{at, unit:u.unit});
    }

    // R4 — DENOMINATOR HAS A BASIS.
    // A catalogue basis must name the catalogue by digest; an enumeration must name
    // its method; a declared basis is admissible but must be marked as the weakest.
    const b = s.basis ?? {};
    if (!BASIS.has(b.kind)) F('R4-denominator-basis','basis.kind absent or outside vocabulary',{at, kind:b.kind ?? null});
    else if (b.kind === 'catalogue' && !HEX.test(String(b.catalogue_digest ?? '')))
      F('R4-denominator-basis','catalogue basis without a catalogue digest',{at});
    else if (b.kind === 'enumeration' && !String(b.enumeration_method ?? '').length)
      F('R4-denominator-basis','enumeration basis without a stated method',{at});

    // R5 — COUNTS ARE NON-NEGATIVE AND EXAMINED CANNOT EXCEED ELIGIBLE.
    if (!(Number.isInteger(s.eligible) && Number.isInteger(s.examined)) || s.eligible < 0 || s.examined < 0)
      F('R5-counts-well-formed','eligible or examined is not a non-negative integer',{at});
    else if (s.examined > s.eligible)
      F('R5-counts-well-formed','examined exceeds eligible',{at, eligible:s.eligible, examined:s.examined});
  }

  // R6 — ABSENCE IS SCOPED.
  // An assertion that something was not observed, made without naming the population
  // it was not observed in, is unfalsifiable and is refused.
  for (const a of (doc.absence_assertions ?? [])) {
    if (!a.stratum || !ids.has(a.stratum))
      F('R6-absence-is-scoped','absence assertion names no stratum, or an unknown one',
        {assertion:String(a.assertion ?? '').slice(0,90), stratum:a.stratum ?? null});
  }

  // R7 — INCOMPLETE EXECUTION MAY NOT BE REPORTED AS CLEAN.
  // If any dispatched unit failed, was exhausted, or could not be dispatched, the
  // run did not complete, and a verdict resting on it must be capped and say so.
  const hardStops = doc.strata.flatMap(s => (s.unexamined ?? [])
    .filter(u => u.disposition === 'failed' || u.disposition === 'resource_exhausted' || u.disposition === 'unavailable')
    .map(u => `${s.id}:${u.unit}:${u.disposition}`));
  if (hardStops.length && doc.integrity.complete === true)
    F('R7-incomplete-not-clean','integrity.complete is true while units failed, exhausted or were unavailable',
      {stops:hardStops.slice(0,6), count:hardStops.length});
  if (doc.integrity.complete === false && !String(doc.integrity.capped_to ?? '').length)
    F('R7-incomplete-not-clean','integrity.complete is false and no capped_to verdict is stated',{});

  // R8 — SUPPORTS IS A BOUND, NOT A DECORATION.
  // Where a stratum is cited by an absence assertion it must state what claim classes
  // its coverage supports, so a reader can see whether the citation is in bounds.
  const cited = new Set((doc.absence_assertions ?? []).map(a => a.stratum));
  for (const s of doc.strata) {
    if (cited.has(s.id) && !(Array.isArray(s.supports) && s.supports.length))
      F('R8-supports-bounds-citation','stratum is cited by an absence assertion but states no supported claim classes',{at:`strata[${s.id}]`});
  }

  return { ok: fail.length === 0, failures: fail };
}

export function summarise(doc) {
  const rows = (doc.strata ?? []).map(s => {
    const by = {};
    for (const u of (s.unexamined ?? [])) by[u.disposition] = (by[u.disposition] ?? 0) + 1;
    return { id:s.id, population:s.population, eligible:s.eligible, examined:s.examined,
             fraction: s.eligible ? (s.examined / s.eligible) : 0, dispositions: by };
  });
  return rows;
}

// CLI
if (process.argv[1] && process.argv[1].endsWith('verify.mjs')) {
  const path = process.argv[2];
  if (!path) { console.error('usage: node verify.mjs <cap-1.json>'); process.exit(2); }
  let doc; try { doc = JSON.parse(readFileSync(path, 'utf8')); }
  catch (e) { console.error('REFUSED  unreadable or invalid JSON:', e.message); process.exit(1); }
  const r = verify(doc);
  if (r.ok) {
    console.log('CONFORMS  cap/1');
    for (const s of summarise(doc)) {
      const pct = (100 * s.fraction).toFixed(1).padStart(5);
      const d = Object.entries(s.dispositions).map(([k,v]) => `${k}=${v}`).join('  ') || 'none';
      console.log(`  ${String(s.id).padEnd(14)} ${String(s.examined).padStart(6)} / ${String(s.eligible).padEnd(6)} ${pct}%   ${d}`);
    }
    process.exit(0);
  }
  console.log('REFUSED  cap/1');
  for (const f of r.failures) console.log(`  ${f.rule}  ${f.grounds}\n      ${JSON.stringify(f.evidence ?? {})}`);
  process.exit(1);
}

File: cap-1/src/verify.py
#!/usr/bin/env python3
# =============================================================================
# CAP-1 CONFORMANCE VERIFIER — SECOND IMPLEMENTATION
#
# Written from the prose of the eight normative rules in
# CS-SPEC-CAP-1-Coverage-Attestation-Profile, not ported from verify.mjs.
# Standard library only. No dependencies, no network, no clock.
#
# WHAT THIS CLOSES AND WHAT IT DOES NOT.
#   Closes: IMPLEMENTATION independence. Two implementations, two languages,
#           two authors' readings of one text, agreeing on every vector.
#   Does NOT close: AUTHOR independence. Both implementations and the vectors
#           originate with the same party, so a misreading shared between them
#           is invisible to this check. The class becomes genuinely two-sided
#           the first time it refuses a document written elsewhere.
#
# Where this implementation had to make a judgement the text did not settle,
# the judgement is recorded in a DIVERGENCE NOTE beside the rule. Those notes
# are the specification defects this exercise exists to surface.
# =============================================================================
import json, re, sys

DISPOSITIONS = {
    "not_applicable", "disabled_by_policy", "unsupported_input",
    "resource_exhausted", "failed", "unavailable", "out_of_scope", "withheld",
}
BASIS_KINDS = {"catalogue", "enumeration", "declared"}
HEX = re.compile(r"^[0-9a-f]{32,128}$")

# A hard stop is a disposition that means the unit was dispatched or should have
# been, and did not produce a result. R7 turns on this set.
# DIVERGENCE NOTE (R7): the text says "failed, exhausted or was unavailable".
# It does not say whether out_of_scope or disabled_by_policy count. Read
# narrowly: both are deliberate, accounted decisions, not incomplete execution.
HARD_STOPS = {"failed", "resource_exhausted", "unavailable"}


def verify(doc):
    """Return (ok, failures). Each failure is (rule, grounds, evidence)."""
    f = []
    def fail(rule, grounds, evidence=None):
        f.append({"rule": rule, "grounds": grounds, "evidence": evidence or {}})

    # R0 — shape. Refuse rather than coerce.
    if not isinstance(doc, dict):
        return False, [{"rule": "R0-shape", "grounds": "not an object", "evidence": {}}]
    if doc.get("profile") != "cap/1":
        fail("R0-shape", "profile is not cap/1", {"profile": doc.get("profile")})
    subj = doc.get("subject")
    if not isinstance(subj, dict) or not isinstance(subj.get("ref"), str):
        fail("R0-shape", "subject.ref absent", {})
    strata = doc.get("strata")
    if not isinstance(strata, list) or not strata:
        fail("R0-shape", "no strata", {})
    integ = doc.get("integrity")
    if not isinstance(integ, dict) or not isinstance(integ.get("complete"), bool):
        fail("R0-shape", "integrity.complete absent", {})
    if f:
        return False, f

    ids = set()
    for s in strata:
        at = "strata[%s]" % s.get("id", "?")

        sid = s.get("id")
        if not isinstance(sid, str) or sid in ids:
            fail("R0-shape", "stratum id absent or duplicated", {"at": at})
        ids.add(sid)

        # R1 — no silent remainder.
        unex = s.get("unexamined")
        if not isinstance(unex, list):
            fail("R1-no-silent-remainder", "unexamined is not an array", {"at": at})
            continue
        eligible, examined = s.get("eligible"), s.get("examined")
        if isinstance(eligible, int) and isinstance(examined, int):
            if examined + len(unex) != eligible:
                fail("R1-no-silent-remainder",
                     "eligible does not equal examined plus accounted unexamined",
                     {"at": at, "eligible": eligible, "examined": examined,
                      "accounted": len(unex), "remainder": eligible - examined - len(unex)})

        for u in unex:
            # R2 — closed disposition vocabulary, and every unit names itself.
            if u.get("disposition") not in DISPOSITIONS:
                fail("R2-closed-disposition", "disposition outside the closed vocabulary",
                     {"at": at, "unit": u.get("unit"), "disposition": u.get("disposition")})
            if not isinstance(u.get("unit"), str) or not u.get("unit"):
                fail("R2-closed-disposition", "unexamined entry names no unit", {"at": at})
            # R3 — withholding is digest-bound.
            if u.get("disposition") == "withheld" and not HEX.match(str(u.get("withheld_digest", ""))):
                fail("R3-withholding-digest-bound",
                     "withheld unit carries no digest of the withheld value",
                     {"at": at, "unit": u.get("unit")})

        # R4 — the denominator has a basis.
        b = s.get("basis") or {}
        kind = b.get("kind")
        if kind not in BASIS_KINDS:
            fail("R4-denominator-basis", "basis.kind absent or outside vocabulary",
                 {"at": at, "kind": kind})
        elif kind == "catalogue" and not HEX.match(str(b.get("catalogue_digest", ""))):
            fail("R4-denominator-basis", "catalogue basis without a catalogue digest", {"at": at})
        elif kind == "enumeration" and not str(b.get("enumeration_method", "")):
            fail("R4-denominator-basis", "enumeration basis without a stated method", {"at": at})

        # R5 — counts well formed.
        if not (isinstance(eligible, int) and isinstance(examined, int)) \
           or isinstance(eligible, bool) or isinstance(examined, bool) \
           or eligible < 0 or examined < 0:
            fail("R5-counts-well-formed",
                 "eligible or examined is not a non-negative integer", {"at": at})
        elif examined > eligible:
            fail("R5-counts-well-formed", "examined exceeds eligible",
                 {"at": at, "eligible": eligible, "examined": examined})

    # R6 — absence is scoped.
    for a in doc.get("absence_assertions") or []:
        if not a.get("stratum") or a.get("stratum") not in ids:
            fail("R6-absence-is-scoped",
                 "absence assertion names no stratum, or an unknown one",
                 {"assertion": str(a.get("assertion", ""))[:90], "stratum": a.get("stratum")})

    # R7 — incomplete execution may not be reported as clean.
    stops = ["%s:%s:%s" % (s.get("id"), u.get("unit"), u.get("disposition"))
             for s in strata for u in (s.get("unexamined") or [])
             if u.get("disposition") in HARD_STOPS]
    if stops and integ.get("complete") is True:
        fail("R7-incomplete-not-clean",
             "integrity.complete is true while units failed, exhausted or were unavailable",
             {"stops": stops[:6], "count": len(stops)})
    if integ.get("complete") is False and not str(integ.get("capped_to") or ""):
        fail("R7-incomplete-not-clean",
             "integrity.complete is false and no capped_to verdict is stated", {})

    # R8 — supports bounds citation.
    cited = {a.get("stratum") for a in (doc.get("absence_assertions") or [])}
    for s in strata:
        if s.get("id") in cited:
            sup = s.get("supports")
            if not (isinstance(sup, list) and sup):
                fail("R8-supports-bounds-citation",
                     "stratum is cited by an absence assertion but states no supported claim classes",
                     {"at": "strata[%s]" % s.get("id")})

    return (len(f) == 0), f


def summarise(doc):
    rows = []
    for s in doc.get("strata") or []:
        by = {}
        for u in s.get("unexamined") or []:
            by[u.get("disposition")] = by.get(u.get("disposition"), 0) + 1
        el = s.get("eligible") or 0
        rows.append({"id": s.get("id"), "population": s.get("population"),
                     "eligible": el, "examined": s.get("examined"),
                     "fraction": (s.get("examined") / el) if el else 0.0,
                     "dispositions": by})
    return rows


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("usage: python3 verify.py <cap-1.json>", file=sys.stderr); sys.exit(2)
    try:
        d = json.load(open(sys.argv[1], encoding="utf-8"))
    except Exception as e:
        print("REFUSED  unreadable or invalid JSON:", e); sys.exit(1)
    ok, fails = verify(d)
    if ok:
        print("CONFORMS  cap/1")
        for s in summarise(d):
            disp = "  ".join("%s=%d" % kv for kv in s["dispositions"].items()) or "none"
            print("  %-14s %6d / %-6d %5.1f%%   %s"
                  % (s["id"], s["examined"], s["eligible"], 100 * s["fraction"], disp))
        sys.exit(0)
    print("REFUSED  cap/1")
    for x in fails:
        print("  %s  %s\n      %s" % (x["rule"], x["grounds"], json.dumps(x["evidence"])))
    sys.exit(1)
